### PR TITLE
Add fun Operation.adapter()

### PR DESCRIPTION
**Pull request by martinbonnin**
This PR is based on https://github.com/apollographql/apollo-android/pull/2830 so the diff is slightly bigger than what it should be.

It removes `ResponseFieldMapper` and uses `ResponseAdapter` instead.

~Edit: somehow Github didn't start actions here 🤷  so the build status and conflicts below are mostly out of date~ Github picked up this PR now

### CODE REVIEW

File: apollo-integration/src/test/kotlin/com/apollographql/apollo/NormalizedCacheTestCase.kt [hunk 1]
```diff
@@ -353,7 +356,7 @@ class NormalizedCacheTestCase {
         Predicate<Response<HeroAndFriendsNamesWithIDsQuery.Data>> { response -> !response.hasErrors() }
     )
     val heroWithFriendsFragment = apolloClient.apolloStore.read(
-        HeroWithFriendsFragment.Mapper(),
+        HeroWithFriendsFragmentImpl_ResponseAdapter,
```
Comment (martinbonnin) [inline apollo-integration/src/test/kotlin/com/apollographql/apollo/NormalizedCacheTestCase.kt:359]: I like that using the imperative store API forces to reference the fragment implementation.

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/Adaptable.kt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.apollographql.apollo.api
+
+import com.apollographql.apollo.api.internal.ResponseAdapter
+
+interface Adaptable<T> {
+  fun adapter(): ResponseAdapter<T>
```
Comment (martinbonnin) [inline apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/Adaptable.kt:6]: A new interface for classes that can `Adapt` others. I'm not 100% convinced on the naming but there is some potential to uniformize a lot of the code in `RealApolloStore` for operation/fragments

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/GraphqlFragment.kt [hunk 3]
```diff
@@ -1,14 +1,3 @@
...
- * Represents a GraphQL fragment
- */
-interface GraphqlFragment {
-
-  /**
-   * Returns marshaller [ResponseFieldMarshaller] to serialize fragment data
-   */
-  fun marshaller(): ResponseFieldMarshaller
-}
+interface GraphqlFragment
```
Comment (sav007) [inline apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/GraphqlFragment.kt:3]: So fragment is no longer `Adaptable `?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/codegen/FragmentType.kt [hunk 4]
```diff
@@ -81,6 +67,14 @@ internal fun CodeGenerationAst.FragmentType.implementationTypeSpec(generateAsInt
       .typeSpec()
       .toBuilder()
       .addSuperinterface(GraphqlFragment::class)
+      .addSuperinterface(Adaptable::class.asTypeName().parameterizedBy(defaultImplementationType.typeRef.asTypeName()))
+      .addFunction(
+          FunSpec.builder("adapter")
+              .addModifiers(KModifier.OVERRIDE)
+              .returns(ResponseAdapter::class.asTypeName().parameterizedBy(defaultImplementationType.typeRef.asTypeName()))
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/codegen/FragmentType.kt:74]: nit: 
`this.defaultImplementationType.typeRef.asAdapterTypeName()`

File: apollo-normalized-cache/src/jvmMain/kotlin/com/apollographql/apollo/cache/normalized/ApolloStore.kt [hunk 5]
```diff
@@ -138,22 +137,20 @@ interface ApolloStore {
...
    *
-   * @param fieldMapper [ResponseFieldMapper] to be used for field mapping
    * @param cacheKey    [CacheKey] to be used to find cache record for the fragment
    * @param variables   [Operation.Variables] required for fragment arguments resolving
    * @param <F>         type of fragment to be read
    * @return {@ApolloStoreOperation} to be performed, that will be resolved with cached fragment data
   </F> */
-  fun <F : GraphqlFragment> read(
-      fieldMapper: ResponseFieldMapper<F>,
+  fun <F> read(
```
Comment (sav007) [inline apollo-normalized-cache/src/jvmMain/kotlin/com/apollographql/apollo/cache/normalized/ApolloStore.kt:152]: we should probably leave only one API (remove `fun <D : Operation.Data, V : Operation.Variables> read(...)`

File: apollo-normalized-cache/src/jvmMain/kotlin/com/apollographql/apollo/cache/normalized/ApolloStore.kt [hunk 6]
```diff
@@ -199,7 +196,7 @@ interface ApolloStore {
    * have changed
    */
   fun write(
```
Comment (sav007) [inline apollo-normalized-cache/src/jvmMain/kotlin/com/apollographql/apollo/cache/normalized/ApolloStore.kt:198]: same here I think we can leave only one `write`. Also for symmetry should it be
```
fun <T> write(data:T, adapter: ResponseAdapter<T>...)
```

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/GraphqlFragment.kt (hunk 3 above)
Comment (sav007) [inline apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/GraphqlFragment.kt:3]: nit: keep KDoc?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/codegen/FragmentType.kt (hunk 4 above)
Comment (martinbonnin) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/codegen/FragmentType.kt]: This is actually not the same:

1.
```
override fun adapter(): ResponseAdapter<Data>
```

2.
```
override fun adapter(): TestQuery_ResponseAdapter
```

I think 1. is cleaner in the generated model, the `_ResponseAdapter` file being an implementation detail?

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/GraphqlFragment.kt (hunk 3 above)
Comment (martinbonnin) [reply apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/GraphqlFragment.kt:3]: It's not that helpful but it does make the file nicer :D ! Overall, I'd like to put a more descriptive KDoc but I'm not sure what to put there. It actually feels like this interface isn't really useful anymore...

**Review by sav007 — APPROVED** [review]
LGTM

File: apollo-normalized-cache/src/jvmMain/kotlin/com/apollographql/apollo/cache/normalized/ApolloStore.kt (hunk 5 above)
Comment (martinbonnin) [reply apollo-normalized-cache/src/jvmMain/kotlin/com/apollographql/apollo/cache/normalized/ApolloStore.kt:152]: Follow up PR for store APIs: https://github.com/apollographql/apollo-android/pull/2839

**Pull request merged**